extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java

```diff
@@ -0,0 +1,36 @@
+package io.quarkus.agroal.runtime.health;
+
+import java.sql.SQLException;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+import javax.sql.DataSource;
+
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.HealthCheckResponse;
+import org.eclipse.microprofile.health.HealthCheckResponseBuilder;
+import org.eclipse.microprofile.health.Readiness;
+
+@Readiness
+@ApplicationScoped
+public class DataSourceHealthCheck implements HealthCheck {
+    @Inject
+    Instance<DataSource> dataSources;
+
+    @Override
+    public HealthCheckResponse call() {
+        HealthCheckResponseBuilder builder = HealthCheckResponse.named("Database connection(s) health check").up();
+        for (DataSource dataSource : dataSources) {
+            try {
+                boolean valid = dataSource.getConnection().isValid(0);
+                if (!valid) {
+                    builder.down().withData("reason", "validation check failed");
```

Comment:
I think this will be overridden if two datasources are down? 
Since we support multiple datasources, will it be possible to return an object having this structure in case of failures? 
``` 
{
 "reason": {
    "validation check failed",
     "datasources": ["data-source-name-1", "data-source-name-2", ..., "data-source-name-N"]
  }
} 
```

Reply:
In case multiple datasources down, there will be multiple data with the same content as the `withData` method adds the data to a HashMap.

First I want to implement exactly what you ask ... then I realize that I don't know how to retrieve the beans with there name in Arc :) 

In case multiple beans of the same type I assume they will have different qualifier. But in CDI to retrieve multiple beans of the same type you use `Instance<Datasource>` that provides the list of the beans and not there names ... If someone can explain me how to retrieve the list of beans from ArC with there qualifier (name) I'll update my implementation.

Reply:
Totaly agree but as I reply to @machi1990 just above, I don't know how to retrieve the bean names from ArC, I use `@Inject Instance<DataSource> dataSources;` that gives me all datasources but not the bean name/qualifier ...

Reply:
> ... If someone can explain me how to retrieve the list of beans from ArC with there qualifier (name) I'll update my implementation.

@loicmathieu I am certainly not a CDI expert, but have you tried something like this? 
```
 final Set<Bean<?>> beans = Arc.container().beanManager().getBeans(javax.sql.DataSource.class);
        for (Bean<?> bean : beans) {
            if (bean.getName() == null) {
                //  default DataSource has not name can be retrieved via
                //  @Inject
                //  javax.sql.DataSource defaultDataSource;

                //TODO do something with defaultDataSource
            } else {
                InstanceHandle<javax.sql.DataSource> instanceHandle = Arc.container().instance(bean.getName());
                if (instanceHandle.isAvailable()) {
                    javax.sql.DataSource dataSource = instanceHandle.get();
                    // TODO do something with dataSource
                }
            }
        }
```
Also pinging @mkouba @manovotn, may be there is a better way to achieve this  :-)

Reply:
Your solution is viable, otherwise you could use `@Any Instance<DataSource>` which would bring you all beans of type `DataSource` regardless of their qualifiers. You can then iterate over all of them and inspect them.

If you are after bean names (and not `DataSource` instances), then going via `Arc.container().beanManager().getBeans(...)` is better.

Reply:
Thanks @machi1990 and @manovotn I used `Arc.container().beanManager().getBeans(...)` in a `  @PostConstruct` and it works fine!

Reply:
@kenfinnigan the first argument of `withData` should be a valid json attribute, for the moment I ended up generating `reason1`, `reason2`, ... but maybe `datasourceName-check-status` will be better but the default datasource has no name and if I put `default` it can collide with someone that use a named datasource of name `default`...

Reply:
No, it surprises me so I check but you can have a default datasource (with no name) and a nammed datasource. See [Multiple Datasource](https://quarkus.io/guides/datasource-guide#multiple-datasources).

So I can configure something like this:
```
# default un-named datasource : can be injected with @Inject with no other annotation
quarkus.datasource.driver=org.h2.Driver
quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:default
quarkus.datasource.username=username-default
quarkus.datasource.min-size=3
quarkus.datasource.max-size=13

# Named datasource with the name default, must be injected with @Datasource annotation
quarkus.datasource.default.driver=org.h2.Driver
quarkus.datasource.default.url=jdbc:h2:tcp://localhost/mem:users
quarkus.datasource.default.username=username1
quarkus.datasource.default.min-size=1
quarkus.datasource.default.max-size=11
```

Reply:
It's just a bit misleading for the developer to talk about the datasource named default in this case: do it refers to the default unamed datasource or the named default datasource ? My proposal clearly talk about the default datasource or the datasource named 'default' in this case.

Reply:
You can, the default datasource has no name (name = null)